### ISSUE TITLE
Question: S1168 reported if returning class implements IEnumerable

### ISSUE THREAD

**Issue by lg2de** [SonarSource/sonar-dotnet#3056]
### Description

If found code issues S1168 "Empty arrays and collections should be returned instead of null" on methods returning class which implement IEnumerable. This class is not primary intended as enumerable, but children should be accessible as enumerable. For these classes `null` is really different from Enumerable.Empty.

I'm unsure whether this is a false positive or code smell in class design.

I tried to understand the current logic of S1168 and identified the following summary:
_All types inherit from `IEnumerable` (or `Array`) should never return `null`. There are only two exception: `string` and `Xml*`._

**Question**: Why XmlNode, etc. are ignored while other classes implementing IEnumerable are not ignored?
Unfortunately I did not find any detailed discussion in #1037.

### Repro steps

```cs
public ImplementingEnumerable GetImplementingEnumerable()
{
    return null; // Compliant classes implementing IEnumerable are ignored
}

public class ImplementingEnumerable : IEnumerable//<int>
{
    //public IEnumerator<int> GetEnumerator() => throw new System.NotImplementedException();
    IEnumerator IEnumerable.GetEnumerator() => throw new System.NotImplementedException();
};
```

### Related information

* SonarC# Version: 8.2
* Visual Studio Version: 16.4.3

**Comment by andrei-epure-sonarsource** [comment]
hi @lg2de 

Regarding https://jira.sonarsource.com/browse/RSPEC-1168

> This class is not primary intended as enumerable, but children should be accessible as enumerable. 

Indeed, it's the same case as XmlNode - [XmlNode.GetEnumerator](https://docs.microsoft.com/en-us/dotnet/api/system.xml.xmlnode.getenumerator?view=netframework-4.8#System_Xml_XmlNode_GetEnumerator) _Gets an enumerator that iterates through the child nodes in the current node._ , so for a method returning XmlNode, it doesn't make sense to return an empty enumeration.

> I'm unsure whether this is a false positive or code smell in class design.

From my point of view, it is a false positive.

Is it a design smell? I think it is a bit, in the sense that for example `XNode` is not a collection class.

If you read [the docs for IEnumerable](https://docs.microsoft.com/en-us/dotnet/api/system.collections.ienumerable?view=netframework-4.8), it mentions its usage only for collection classes.

To reduce the noise of this rule, it could raise only when the returned type is `IEnumerable`, `ICollection` or a known collection type, but this would come with a lot of false negatives.

`XNode` is part of a very used API and this is why it was added as an exception...

WDYT?

**Comment by lg2de** [comment]
> To reduce the noise of this rule, it could raise only when the returned type is `IEnumerable`, `ICollection` or a known collection type, but this would come with a lot of false negatives.

What kind of "false negatives" do you see?

The rule should apply to types designed to _be enumerable only_.
How to identify these types?
As soon as the type _has members of any type_ there **is** a difference between `null` and "default constructed".
Can we check for all members?

**Comment by andrei-epure-sonarsource** [comment]
> As soon as the type has members of any type there is a difference between null and "default constructed".
> Can we check for all members?

By members I assume you mean properties (we exclude methods and indexers). Let's exclude primitive types (e.g. `Count`, `IsReadOnly`).

But how can you tell if some class is a "true" collection or not?
- `LinkedList<T>` has a [First](https://docs.microsoft.com/en-us/dotnet/api/system.collections.generic.linkedlist-1.first?view=netframework-4.8#System_Collections_Generic_LinkedList_1_First) property which is of type `LinkedListNode<T>`
- `SortedList` has `SortedList.Keys` and `SortedList.Values`, which are both `ICollection`

IMO `XNode` has an unusual design. The [`IEnumerable` interface](https://docs.microsoft.com/en-us/dotnet/api/system.collections.ienumerable?view=netframework-4.8) says:

> Exposes an enumerator, which supports a simple iteration over a non-generic collection.

If there's a mix of concerns inside a class, that's a design problem. In the case of `XmlNode`, normally it should just have a `Children` property that is enumerable.

I am closing this issue as I don't see a clear way of solving this corner case. WDYT?
